adapters/repos/db/vector/cuvs/index.go

```diff
@@ -0,0 +1,739 @@
+	}
+
+	logger := cfg.Logger
+	if logger == nil {
+		l := logrus.New()
+		l.Out = io.Discard
+		logger = l
+	}
+
+	res, err := cuvs.NewResource(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.CuvsIndexParams == nil {
+		cuvsIndexParams, err := cagra.CreateIndexParams()
+
+		cuvsIndexParams.SetGraphDegree(32)
+		cuvsIndexParams.SetIntermediateGraphDegree(32)
+		cuvsIndexParams.SetBuildAlgo(cagra.NnDescent)
+
+		if err != nil {
+			return nil, err
+		}
+
+		cfg.CuvsIndexParams = cuvsIndexParams
+
+	}
+
+	if cfg.CuvsSearchParams == nil {
+
+		cuvsSearchParams, err := cagra.CreateSearchParams()
+		if err != nil {
+			return nil, err
+		}
+
+		cuvsSearchParams.SetAlgo(cagra.SearchAlgoMultiCta)
+		cuvsSearchParams.SetItopkSize(256)
+		cuvsSearchParams.SetSearchWidth(1)
+
+		cfg.CuvsSearchParams = cuvsSearchParams
+
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	cuvsIndex, err := cagra.CreateIndex()
+	if err != nil {
+		return nil, fmt.Errorf("create cuvs index: %w", err)
+	}
+
+	internals := cuvs_internals{
+		cuvsIndex:        cuvsIndex,
+		cuvsIndexParams:  cfg.CuvsIndexParams,
+		cuvsSearchParams: cfg.CuvsSearchParams,
+		cuvsResource:     &res,
+		dlpackTensor:     nil,
+	}
+
+	index := &cuvs_index{
+		cuvs_internals: internals,
+		id:             cfg.ID,
+		targetVector:   cfg.TargetVector,
+		logger:         logger,
+		distanceMetric: cuvs.DistanceL2, // TODO: make configurable
+		store:          store,
+		rootPath:       cfg.RootPath,
+
+		idCuvsIdMap:    make(map[uint32]uint64),
+		cuvsPoolMemory: cfg.CuvsPoolMemory,
+	}
+
+	println("pre init buckets")
+
+	if err := index.initBuckets(context.Background()); err != nil {
+		return nil, fmt.Errorf("init cuvs index buckets: %w", err)
+	}
+	println("post init buckets")
+
+	if err := index.initMetadata(); err != nil {
+		return nil, err
+	}
+
+	return index, nil
+}
+
+func byteSliceFromFloat32Slice(vector []float32, slice []byte) []byte {
+	for i := range vector {
+		binary.LittleEndian.PutUint32(slice[i*4:], math.Float32bits(vector[i]))
+	}
+	return slice
+}
+
+func shouldExtend(index *cuvs_index, num_new uint64) bool {
+	if num_new > 100 {
+		// println("num_new is greater than 100; rebuilding")
+		return false
+	}
+
+	if index.count < 300_000 {
+		// println("index count is less than 300_000; rebuilding")
+		return false
+	}
+
+	if index.dlpackTensor == nil {
+		// println("dlpack tensor is nil; rebuilding")
+		return false
+	}
+
+	percentNewVectors := (float32(num_new+index.cuvsExtendCount) / float32(index.count+num_new)) * 100
+
+	// why 20? https://weaviate-org.slack.com/archives/C05V3MGDGQY/p1722897390825229?thread_ts=1722894509.398509&cid=C05V3MGDGQY
+	if percentNewVectors > 20 {
+		println("percentNewVectors is greater than 20; rebuilding")
+		return false
+	}
+
+	return true
+}
+
+func Normalize_Temp(vector []float32) []float32 {
+	newVector := make([]float32, len(vector))
+	for i := range vector {
+		newVector[i] = vector[i] * 0.146
```

Comment:
What does this function do? The `0.146` value doesn't look like it will work for different dimensions sizes.

Reply:
This is a workaround to a bug in the cuvs extend feature. I've spoken to nvidia about this already and will make an issue